streams/shared/src/main/scala/zio/stream/experimental/ZSink.scala

```diff
@@ -215,6 +215,97 @@ abstract class ZSink[-R, +E, -I, +Z] private (
       }
     }
 
+  /**
+   * A named alias for `race`.
+   */
+  final def |[R1 <: R, E1 >: E, A0, I1 <: I, Z1 >: Z](
+    that: ZSink[R1, E1, I1, Z1]
+  ): ZSink[R1, E1, I1, Z1] =
+    self.race(that)
+
+  /**
+   * Runs both sinks in parallel on the input, returning the result from the
+   * one that finishes successfully first.
+   */
+  final def race[R1 <: R, E1 >: E, A0, I1 <: I, Z1 >: Z](
+    that: ZSink[R1, E1, I1, Z1]
+  ): ZSink[R1, E1, I1, Z1] =
+    self.raceBoth(that).map(_.merge)
+
+  /**
+   * Runs both sinks in parallel on the input, returning the result from the
+   * one that finishes successfully first.
+   */
+  final def raceBoth[R1 <: R, E1 >: E, A0, I1 <: I, Z1](
+    that: ZSink[R1, E1, I1, Z1]
+  ): ZSink[R1, E1, I1, Either[Z, Z1]] = {
+
+    case class SingleSinkState(active: Push[R1, E1, I1, Any], isFirst: Boolean, otherSinkFailure: E1) {
+      def halt(err: E1): IO[Either[E1, Nothing], Nothing] = {
+        val (e1, e2) = if (isFirst) (err, otherSinkFailure) else (otherSinkFailure, err)
+        Push.halt(Cause.Both(Cause.fail(e1), Cause.fail(e2)))
+      }
+
+      def emit(z: Any): ZIO[Any, Either[Nothing, Either[Z, Z1]], Nothing] =
+        if (isFirst) Push.emit(Left(z.asInstanceOf[Z])) else Push.emit(Right(z.asInstanceOf[Z1]))
+    }
+
+    ZSink(for {
+      ref <- ZRef.makeManaged[Option[SingleSinkState]](None)
+      p1  <- self.push
+      p2  <- that.push
+      push: Push[R1, E1, I1, Either[Z, Z1]] = {
+        in =>
+          ref.get.flatMap {
+            case Some(onlyOne) => {
+              onlyOne
+                .active(in)
+                .foldM({
+                  case Left(err) => onlyOne.halt(err)
+                  case Right(z)  => onlyOne.emit(z)
+                }, _ => Push.more)
+            }
+            case None => {
+              p1(in).either.zipPar(p2(in).either).flatMap {
+                case (res1, res2) => {
```

Comment:
![image](https://user-images.githubusercontent.com/283332/79161117-9ccdc280-7de3-11ea-83e0-672082ec7ef8.png)

Reply:
Thanks god there are only 2 sinks to race!
I tried to come with a state structure that could handle 2 updates one-by-one.
But after few attempts just threw a pragmatic hadouken.